### DISCUSSION TITLE
Order of validating elements of union type

### DISCUSSION THREAD

**Discussion by dslatkin** [colinhacks/zod#1828] [Q&A] [answered]
Hey all,

I'm a bit stuck on this and think maybe my mental model of Zod is wrong and I'd love some help to figure out why. I'm working with an `<input type="date" />` element and parsing the `FormData` it sends to the server. There are two cases I'd like handled:

1. If the user doesn't select a date, the empty string `""` gets sent and I'd like it to get transformed to `undefined`.
2. If the user does select a date, a string like `"2023-01-01"` gets sent and I'd like it to get coerced into a `Date` object and refined to check that it's in the past, like for a birthday.

If I do it in this order, things work as I would expect for all three cases (empty string and date in past pass, date in future fails):
```ts
z.union([
    z.coerce.date().refine(
        (d) => {
            return d.getTime() < Date.now();
        },
        { message: 'Birthday must be in the past' },
    ),
    z
        .string()
        .length(0)
        .transform(() => undefined),
])
```

But if I try it in this order:
```ts
z.union([
    z
        .string()
        .length(0)
        .transform(() => undefined),
    z.coerce.date().refine(
        (d) => {
            return d.getTime() < Date.now();
        },
        { message: 'Birthday must be in the past' },
    ),
])
```
and I select a date in the future, for some reason I'm trying to figure out, I get the error "String must contain exactly 0 character(s)".

My understanding from the README is that the schemas get tested in order and the first to parse successfully gets returned. How then, I'm wondering, is it failing on a refine in the second schema, but then returning an error coming from the first schema?

All help is appreciated, I feel like I must be missing something obvious. Thank you!

**Comment by JacobWeisenburger** [comment, marked as answer]
I have confirmed this. It seems to be something about the refinement that causes the union to use the other option. Not exactly sure what's going on here. But I'm pretty sure this is a bug.

```ts
{
    const result = z.union( [
        z.coerce.date().refine(
            d => d.getTime() < Date.now(),
            { message: 'Birthday must be in the past' },
        ),
        z.string().length( 0 ),
    ] ).safeParse( '2024-01-01' )

    result.success
        ? console.log( result.data )
        : console.log( result.error.issues[ 0 ].message )
    // logs: Birthday must be in the past
    // works as expected
}

{
    const result = z.union( [
        z.string().length( 0 ),
        z.coerce.date().refine(
            d => d.getTime() < Date.now(),
            { message: 'Birthday must be in the past' },
        ),
    ] ).safeParse( '2024-01-01' )

    result.success
        ? console.log( result.data )
        : console.log( result.error.issues[ 0 ].message )
    // logs: String must contain exactly 0 character(s)
    // does not work as expected
    // expected: Birthday must be in the past
}

{
    const result = z.union( [
        z.string().length( 0 ),
        z.coerce.date(),
    ] ).safeParse( '2024-01-01' )

    result.success
        ? console.log( result.data )
        : console.log( result.error.issues[ 0 ].message )
    // logs: 2024-01-01T00:00:00.000Z
    // works as expected
}

{
    const result = z.union( [
        z.coerce.date(),
        z.string().length( 0 ),
    ] ).safeParse( '2024-01-01' )

    result.success
        ? console.log( result.data )
        : console.log( result.error.issues[ 0 ].message )
    // logs: 2024-01-01T00:00:00.000Z
    // works as expected
}
```

**Comment by dslatkin** [comment]
Thanks for looking into it for me, I really appreciate it! I'll probably type this up and submit it as an issue next week.

**Reply by JacobWeisenburger** [reply]
sounds good. thanks
